test/test_namespacemanager.py

```diff
@@ -0,0 +1,47 @@
+import sys
+from pathlib import Path
+sys.path.append(str(Path(__file__).parent.parent.absolute()))
+from rdflib import Graph
+from rdflib.namespace import NAMESPACE_PREFIXES_CORE, NAMESPACE_PREFIXES_RDFLIB
+from rdflib.namespace import OWL
+
+
+def test_core_prefixes_bound():
+    # we should have RDF, RDFS, OWL, XSD & XML bound
+    g = Graph()
+
+    # prefixes in Graph
+    assert len(list(g.namespaces())) == len(NAMESPACE_PREFIXES_CORE)
+    pre = sorted([x[0] for x in list(g.namespaces())])
+    assert pre == ["owl", "rdf", "rdfs", "xml", "xsd"]
+
+
+def test_rdflib_prefixes_bound():
+    g = Graph(bind_namespaces="rdflib")
+
+    # the core 5 + the extra 23 namespaces with prefixes
+    assert len(list(g.namespaces())) == \
+           len(NAMESPACE_PREFIXES_CORE) + len(list(NAMESPACE_PREFIXES_RDFLIB))
+
+
+def test_cc_prefixes_bound():
+    pass
+
+
+def test_rebinding():
+    g = Graph()  # 'core' bind_namespaces (default)
+    print()
+    # 'owl' should be bound
+    assert "owl" in [x for x, y in list(g.namespaces())]
+
+    # replace 'owl' with 'sowa'
+    # 'sowa' should be bound
+    # 'owl' should not be bound
+    g.bind("sowa", OWL, replace=True)
```

Comment:
I think this should be testing with `override` instead of `replace`, as the namespace's prefix is being changed to a new prefix

```
        If override, rebind, even if the given namespace is already
        bound to another prefix.
```

as opposed to changing the namespace associated with a prefix, which is what `replace=True` should be for.

```
        If replace, replace any existing prefix with the new namespace
```

Similarly I think the variable you should be passing is override, not replace.

Reply:
Coupla points. 
1. There's an implicit legacy issue in https://github.com/RDFLib/rdflib/blob/6b356a4eacde1280f78b4ed966e8a499a727189a/rdflib/graph.py#L641 where `Graph.__add__()` accumulates namespace bindings which `Graph.__sub__()` ignores.
2. I'll copy [gromgull's observations](https://github.com/RDFLib/rdflib/issues/543) here, for convenience:
> The confusion here is the two arguments `replace` and `override`
> 
> Say they release FOAF 2.0 (Now based on google+? :) ) and you want to change `foaf` to point to `http://xmlns.com/foaf/2.0/`. Up to now you had `foaf=>http://xmlns.com/foaf/0.1` and `foaf2=>http://xmlns.com/foaf/2.0`
> 
> Now you want to "upgrade" to FOAF 2.0 :
> 
> ```py
> g.namespace_manager.bind('foaf', Namespace('http://xmlns.com/foaf/2.0'), .... )
> ```
> 
> Adding:
> 
>     `override` will do what you want, and replace the current binding for foaf to be the new URI.
> 
>     `replace` will replace *any other bindings pointing to the same namespace*, so the 
>      old foaf2 binding will be removed.
> 
> In the FOAF2 example, you probably want both.

and [further](https://github.com/RDFLib/rdflib/issues/543#issuecomment-158754466)

> The problem is (maybe) in the individual implementation, for instance, here:
> 
> https://github.com/RDFLib/rdflib/blob/master/rdflib/plugins/memory.py#L150
> 
> that when replacing, the `uri=>prefix` is updated correctly (i.e. the 
> http://xmlns.com/foaf/2.0 now points to `foaf3`) but the `prefix=>uri` map of course 
> has both. This should be cleaned up if the prefix was already used.
> 
> Also, an unbind method would be handy. Sometimes you want to remove a prefix without providing a new one.

Reply:
@gjhiggins could you perhaps provide a test for what you see as correct behaviour? We could add that test and the resolution for it here before merging. Happy to get it right one time.

Reply:
Sorry to be holding everyone up. I'm trying to get a very clear model of the apparently-intended behaviour but I'm not there yet:
```py
def test_binding_strategy():

    # The confusion here is in the two arguments “override” and “replace” and
    # how they serve two different purposes.

    # 1. Replacing a binding:
    # Say they release FOAF 2.0 and you want to change `foaf` to point to `http://xmlns.com/foaf/2.0/`.

    # Up to now you had `foaf=>http://xmlns.com/foaf/0.1/` and `foaf2=>http://xmlns.com/foaf/2.0/`
    g = Graph(bind_namespaces="none")
    assert len(list(g.namespaces())) == 0

    foaf1_uri = URIRef('http://xmlns.com/foaf/0.1/')
    foaf2_uri = URIRef("http://xmlns.com/foaf/2.0/")

    FOAF1 = Namespace(foaf1_uri)
    FOAF2 = Namespace(foaf2_uri)

    g.bind("foaf", FOAF1)
    g.bind("foaf2", FOAF2)
    assert len(list(g.namespaces())) == 2
    assert list(g.namespaces()) == [('foaf', foaf1_uri), ("foaf2", foaf2_uri)]

    # Now you want to "upgrade" to foaf=>FOAF 2.0.
    g.bind("foaf", FOAF2)

    # But that doesn't happen, instead a different prefix is bound:
    assert list(g.namespaces()) == [("foaf", foaf1_uri), ("foaf1", foaf2_uri)]

    # The prefix "foaf" was already bound to the FOAF1 namespace, so a new
    # prefix for FOAF2 was created.

    # You can achieve the intended result by setting `replace` to `True`:
    g.bind("foaf", FOAF2, replace=True)

    # Because the FOAF2 namespace was rebound to a different prefix, the old 
    # "foaf2" prefix has gone and because the "foaf" prefix was rebound to a
    # different namespace, the old FOAF1 namespace has gone, leaving just:

    assert list(g.namespaces()) == [("foaf", URIRef("http://xmlns.com/foaf/2.0/"))]

    # Now you can re-bind the FOAF1.0 namespace
    g.bind("oldfoaf", FOAF1)

    with pytest.raises(AssertionError):
        assert list(g.namespaces()) == [("foaf", foaf2_uri), ('oldfoaf', foaf1_uri)]

    assert list(g.namespaces()) == [('oldfoaf', foaf1_uri)]
```

I need to work through the code to figure out why the last `g.bind("oldfoaf", FOAF1)` causes the `foaf` binding to be lost.

Reply:
Oookaayy.
> Just one more review by @gjhiggins please, including a response to my comment about providing a test for expected behaviour.

Kinda lost in the weeds here w.r.t. committing to branches/PRs, the `default_prefixes` branch doesn't appear in the list of branches in my GH clone and I don't seem to be able to change that and in my ignorance, I'm reduced to posting a patch:

```diff
--- rdflib/rdflib/plugins/stores/memory.py  2022-04-08 18:28:03.656450160 +0100
+++ rdflib/rdflib/plugins/stores/memory.py  2022-04-08 16:38:54.877989383 +0100
@@ -403,11 +403,19 @@
                             yield triple, self.__contexts(triple)
 
     def bind(self, prefix, namespace, override=True):
-        bound_prefix = self.__prefix.get(namespace)
-        if override and bound_prefix:
-            del self.__namespace[bound_prefix]
-        self.__prefix[namespace] = prefix
-        self.__namespace[prefix] = namespace
+        bound_namespace = self.__namespace.get(prefix)
+        bound_prefix = self.__prefix.get(namespace) or self.__prefix.get(bound_namespace)
+        if override:
+            if bound_prefix:
+                del self.__namespace[bound_prefix]
+            if bound_namespace:
+                del self.__prefix[bound_namespace]
+
+            self.__prefix[namespace] = prefix
+            self.__namespace[prefix] = namespace
+        else:
+            self.__prefix[bound_namespace or namespace] = bound_prefix or prefix
+            self.__namespace[bound_prefix or prefix] = bound_namespace or namespace
 
     def namespace(self, prefix):
         return self.__namespace.get(prefix, None)
```
I'm unsure whether there's a more elegant means of implementing the above change, so I've refrained from propagating it to the other stores until it's been reviewed.

Reply:
Again, apologies for dumping stuff here rather than via a PR to a PR. I worked up an extended set of tests, the first of which (`test_binding_strategy`) I intend to transcribe to doctest format and migrate into the ["Namespaces and bindings" document](https://github.com/RDFLib/rdflib/blob/master/docs/namespaces_and_bindings.rst) (I'll probably do the same with `test_multigraph_bindings` because it's more of an illustration than a test).

```py
import sys
from pathlib import Path
import pytest

from rdflib.term import URIRef
from rdflib import ConjunctiveGraph, Graph, Literal
from rdflib.plugins.stores.memory import Memory
from rdflib.namespace import Namespace, NAMESPACE_PREFIXES_CORE, NAMESPACE_PREFIXES_RDFLIB
from rdflib.namespace import OWL, RDFS
from test.data import tarek, michel, context0, context1, context2

foaf1_uri = URIRef("http://xmlns.com/foaf/0.1/")
foaf2_uri = URIRef("http://xmlns.com/foaf/2.0/")

FOAF1 = Namespace(foaf1_uri)
FOAF2 = Namespace(foaf2_uri)

def test_binding_strategy():

    # The confusion here is in the two arguments “override” and “replace” and
    # how they serve two different purposes --- changing a prefix already bound
    # to a namespace versus changing a namespace already bound to a prefix.

    g = Graph(bind_namespaces="none")
    assert len(list(g.namespaces())) == 0

    # 1. Changing the namespace of an existing namespace=>prefix binding:

    # Say they release FOAF 2.0 and you want to change the binding of
    # `http://xmlns.com/foaf/2.0/` to `foaf`.

    # Up to now you had `http://xmlns.com/foaf/0.1/=>foaf` and `http://xmlns.com/foaf/2.0/=>foaf2`

    # We'll just set up those two bindings ...
    g.bind("foaf", FOAF1)
    g.bind("foaf2", FOAF2)
    assert len(list(g.namespaces())) == 2
    assert list(g.namespaces()) == [
        ('foaf', foaf1_uri),
        ("foaf2", foaf2_uri)
    ]

    # Now you want to "upgrade" to FOAF2=>foaf and try the obvious:
    g.bind("foaf", FOAF2)

    # But that doesn't happen, instead a different prefix, `foaf1` is bound:
    assert list(g.namespaces()) == [
        ("foaf", foaf1_uri),
        ("foaf1", foaf2_uri)
    ]

    # The rationale behind this behaviour is that the prefix "foaf" was already
    # bound to the FOAF1 namespace, so a different prefix for the FOAF2 namespace 
    # was automatically created.

    # You can achieve the intended result by setting `replace` to `True`:
    g.bind("foaf", FOAF2, replace=True)

    # Because the FOAF2 namespace was rebound to a different prefix, the old 
    # "foaf2" prefix has gone and because the "foaf" prefix was rebound to a
    # different namespace, the old FOAF1 namespace has gone, leaving just:

    assert list(g.namespaces()) == [
        ("foaf", foaf2_uri)
    ]

    # Now, if you wish, you can re-bind the FOAF1.0 namespace to a prefix
    # of your choice
    g.bind("oldfoaf", FOAF1)

    assert list(g.namespaces()) == [
        ("foaf", foaf2_uri),
        ('oldfoaf', foaf1_uri)
    ]

    # 2. Changing the prefix of an existing namespace=>prefix binding:

    # The namespace manager preserves the existing bindings from any
    # subsequent unwanted programmatic rebindings such as:
    g.bind("foaf", FOAF1)

    # Which, as things stand, results in:

    assert list(g.namespaces()) == [
        ("foaf", foaf2_uri),
        ('foaf1', foaf1_uri)
    ]

    # In which the attempted change from `oldfoaf` to (the already 
    # bound-to-a different-namespace `foaf`) was intercepted and
    # changed to a non-clashing prefix of `foaf1`

    # So, after undoing the above unwanted rebinding ..
    g.bind("oldfoaf", FOAF1, replace=True)

    # The bindings are again as desired 
    assert list(g.namespaces()) == [
        ("foaf", foaf2_uri),
        ('oldfoaf', foaf1_uri)
    ]

    # Next time through, set override to False
    g.bind("foaf", FOAF1, override=False)

    # And the bindings will remain as desired
    assert list(g.namespaces()) == [
        ("foaf", foaf2_uri),
        ('oldfoaf', foaf1_uri)
    ]

    # 3. Parsing data with prefix=>namespace bindings 
    # Let's see the situation regarding namespace bindings
    # in parsed data - it can be a bit confusing ...

    # Starting with a very likely example where `foaf` is a
    # prefix for `FOAF1`

    data = """\
    @prefix foaf: <http://xmlns.com/foaf/0.1/> .

    <urn:example:a> foaf:name "Bob" ."""

    g.parse(data=data, format="n3")

    # The FOAF2 namespace is already bound to `foaf` so a
    # non-clashing prefix of `foaf1` is rebound to FOAF1 in
    # place of the existing `oldfoaf` prefix

    assert list(g.namespaces()) == [
        ("foaf", foaf2_uri),
        ('foaf1', foaf1_uri)
    ]

    # Yes, namespace bindings in parsed data replace existing
    # bindings (i.e. `oldfoaf` was replaced by `foaf1`), just
    # live with it ...

    # A different example of the same principle where `foaf2`
    # replaces the exsting `foaf`

    data = """\
    @prefix foaf2: <http://xmlns.com/foaf/2.0/> .

    <urn:example:b> foaf2:name "Alice" ."""

    g.parse(data=data, format="n3")

    # The already-bound namespace of `foaf=>FOAF2` is replaced
    # by the `foaf2=>FOAF2` binding specified in the N3 source

    assert list(g.namespaces()) == [
        ('foaf1', foaf1_uri),
        ("foaf2", foaf2_uri),
    ]

    # Where a prefix-namespace binding in the data does
    # not clash with the existing binding ...

    data = """\
    @prefix foaf1: <http://xmlns.com/foaf/0.1/> .

    <urn:example:a> foaf1:name "Bob" ."""

    g.parse(data=data, format="n3")

    # The prefix `foaf1`, already bound to FOAF1 is
    # used

    assert list(g.namespaces()) == [
        ('foaf1', foaf1_uri),
        ("foaf2", foaf2_uri),
    ]

    # Otherwise, the existing prefix binding is replaced

    data = """\
    @prefix foaf: <http://xmlns.com/foaf/0.1/> .

    <urn:example:b> foaf:name "Alice" ."""

    g.parse(data=data, format="n3")

    assert list(g.namespaces()) == [
        ("foaf2", foaf2_uri),
        ('foaf', foaf1_uri),
    ]

    # Prefixes are bound to namespaces which in turn have a URIRef
    # representation - so a different prefix=>namespace binding
    # means a different predicate and thus a different statement:

    expected = """
        @prefix foaf: <http://xmlns.com/foaf/0.1/> .
        @prefix foaf2: <http://xmlns.com/foaf/2.0/> .

        <urn:example:a> foaf:name "Bob" .

        <urn:example:b> foaf:name "Alice" ;
            foaf2:name "Alice" .

        """

    s = g.serialize(format="n3")

    for l in expected.split():
        assert l in s


def test_prefix_alias_disallowed():

    # CANNOT BIND A PREFIX ALIASED TO AN EXISTING BOUND NAMESPACE

    g = Graph(bind_namespaces="none")
    g.bind("owl", OWL)
    assert ('owl', URIRef('http://www.w3.org/2002/07/owl#')) in list(g.namespaces())

    g.bind("wol", OWL, override=False)
    assert ('owl', URIRef('http://www.w3.org/2002/07/owl#')) in list(g.namespaces())
    assert ('wol', URIRef('http://www.w3.org/2002/07/owl#')) not in list(g.namespaces())


def test_rebind_prefix_succeeds():

    # CAN REPLACE A PREFIX-NAMESPACE BINDING

    g = Graph(bind_namespaces="none")
    g.bind("owl", OWL)
    assert ('owl', URIRef('http://www.w3.org/2002/07/owl#')) in list(g.namespaces())

    g.bind("wol", OWL)
    assert ('wol', URIRef('http://www.w3.org/2002/07/owl#')) in list(g.namespaces())
    assert ('owl', URIRef('http://www.w3.org/2002/07/owl#')) not in list(g.namespaces())


def test_parse_rebinds_prefix():

    # PARSED PREFIX-NAMESPACE BINDINGS REPLACE EXISTING BINDINGS

    data = """@prefix friend-of-a-friend: <http://xmlns.com/foaf/0.1/> .

    <urn:example:a> friend-of-a-friend:name "Bob" .

    """

    # Use full set of namespace bindings, including foaf 
    g = Graph(bind_namespaces="none")
    g.bind('foaf', FOAF1)
    assert ('foaf', foaf1_uri) in list(g.namespaces())

    g.parse(data=data, format="n3")

    # foaf no longer in set of namespace bindings
    assert ('foaf', foaf1_uri) not in list(g.namespaces())
    assert ('friend-of-a-friend', foaf1_uri) in list(g.namespaces())


@pytest.mark.xfail(reason="Automatic handling of unknown predicates not automatically registered with namespace manager")
def test_automatic_handling_of_unknown_predicates():

    # AUTOMATIC HANDLING OF UNKNOWN PREDICATES

    """
    Automatic handling of unknown predicates
    -----------------------------------------

    As a programming convenience, a namespace binding is automatically
    created when :class:`rdflib.term.URIRef` predicates are added to the graph.
    """

    g = Graph(bind_namespaces="none")

    g.add((tarek, URIRef('http://xmlns.com/foaf/0.1/name'), Literal("Tarek")))

    assert len(list(g.namespaces())) > 0


def test_automatic_handling_of_unknown_predicates_only_effected_after_serialization():

    g = Graph(bind_namespaces="none")

    g.add((tarek, URIRef('http://xmlns.com/foaf/0.1/name'), Literal("Tarek")))

    assert "@prefix ns1: <http://xmlns.com/foaf/0.1/> ." in g.serialize(format="n3")

    assert len(list(g.namespaces())) > 0
    assert ("ns1", URIRef("http://xmlns.com/foaf/0.1/")) in list(g.namespaces())


def test_multigraph_bindings():
    data = """@prefix friend-of-a-friend: <http://xmlns.com/foaf/0.1/> .

    <urn:example:a> friend-of-a-friend:name "Bob" .

    """

    store = Memory()

    g1 = Graph(store, identifier=context1, bind_namespaces="none")

    g1.bind('foaf', FOAF1)
    assert list(g1.namespaces()) == [
        ('foaf', foaf1_uri)
    ]
    assert list(store.namespaces()) == [
        ('foaf', foaf1_uri)
    ]

    g1.add((tarek, FOAF1.name, Literal("tarek")))

    assert list(store.namespaces()) == [
        ('foaf', foaf1_uri)
    ]

    g2 = Graph(store, identifier=context2, bind_namespaces="none")
    g2.parse(data=data, format="n3")

    # The parser-caused rebind is in the underlying store and all objects
    # that use the store see the changed binding:
    assert list(store.namespaces()) == [
        ('friend-of-a-friend', foaf1_uri)
    ]
    assert list(g1.namespaces()) == [
        ('friend-of-a-friend', foaf1_uri)
    ]

    # Including newly-created objects that use the store
    cg = ConjunctiveGraph(store=store)

    assert ('foaf', foaf1_uri) not in list(cg.namespaces())
    assert ('friend-of-a-friend', foaf1_uri) in list(cg.namespaces())

    assert len(list(g1.namespaces())) == 6
    assert len(list(g2.namespaces())) == 6
    assert len(list(cg.namespaces())) == 6
    assert len(list(store.namespaces())) == 6

    cg.store.add_graph(g1)
    cg.store.add_graph(g2)

    assert "@prefix friend-of-a-friend: <http://xmlns.com/foaf/0.1/>" in cg.serialize(format="n3")

    # In the notation3 format, the statement asserting tarek's name
    # now references the changed prefix:
    assert '<urn:example:tarek> friend-of-a-friend:name "tarek" .' in cg.serialize(format="n3")

    # Add foaf2 binding if not already bound
    cg.bind("foaf2", FOAF2, override=False)
    assert ('foaf2', foaf2_uri) in list(cg.namespaces())

    # Impose foaf binding ... if not already bound
    cg.bind("foaf", FOAF1, override=False)
    assert ('foaf', foaf1_uri) not in list(cg.namespaces())


def test_new_namespace_new_prefix():
    g = Graph(bind_namespaces="none")
    g.bind("foaf", FOAF1)
    assert list(g.namespaces()) == [
        ('foaf', foaf1_uri)
    ]


def test_change_prefix_override_true():
    g = Graph(bind_namespaces="none")

    g.bind("foaf", FOAF1)
    assert list(g.namespaces()) == [
        ('foaf', foaf1_uri)
    ]

    g.bind("oldfoaf", FOAF1)
    # Changed
    assert list(g.namespaces()) == [
        ("oldfoaf", foaf1_uri)
    ]


def test_change_prefix_override_false():
    g = Graph(bind_namespaces="none")

    g.bind("foaf", FOAF1)
    assert list(g.namespaces()) == [
        ('foaf', foaf1_uri)
    ]

    g.bind("oldfoaf", FOAF1, override=False)
    # No change
    assert list(g.namespaces()) == [
        ("foaf", foaf1_uri)
    ]


def test_change_namespace_override_true():
    g = Graph(bind_namespaces="none")

    g.bind("foaf", FOAF1)
    assert list(g.namespaces()) == [
        ('foaf', foaf1_uri)
    ]

    g.bind("foaf", FOAF2)
    # Different prefix used
    assert list(g.namespaces()) == [
        ("foaf", foaf1_uri),
        ("foaf1", foaf2_uri)
    ]


def test_change_namespace_override_false():
    g = Graph(bind_namespaces="none")

    g.bind("foaf", FOAF1)
    assert list(g.namespaces()) == [
        ('foaf', foaf1_uri)
    ]

    # Different namespace so override is irrelevant in this case
    g.bind("foaf", FOAF2, override=False)
    # Different prefix used
    assert list(g.namespaces()) == [
        ("foaf", foaf1_uri),
        ("foaf1", foaf2_uri)
    ]


def test_new_namespace_override_false():
    g = Graph(bind_namespaces="none")

    g.bind("foaf", FOAF2)
    assert list(g.namespaces()) == [
        ('foaf', foaf2_uri)
    ]

    # Namespace not already bound so override is irrelevant in this case
    g.bind("owl", OWL, override=False)
    # Given prefix used
    assert list(g.namespaces()) == [
        ("foaf", foaf2_uri),
        ('owl', URIRef('http://www.w3.org/2002/07/owl#')),
    ]


def test_change_namespace_and_prefix():

    # A more extensive illustration of attempted rebinds of
    # `foaf` being intercepted and changed to a non-clashing
    # prefix of `foafN` (where N is an incrementing integer) ...

    g = Graph(bind_namespaces="none")

    g.bind("foaf", FOAF1)
    assert list(g.namespaces()) == [
        ('foaf', foaf1_uri)
    ]

    g.bind("foaf", FOAF2, replace=True)
    assert list(g.namespaces()) == [
        ("foaf", foaf2_uri)
    ]

    g.bind("foaf1", FOAF1)

    assert list(g.namespaces()) == [
        ("foaf", foaf2_uri),
        ("foaf1", foaf1_uri)
    ]

    foaf3_uri = URIRef("http://xmlns.com/foaf/3.0/")
    FOAF3 = Namespace("http://xmlns.com/foaf/3.0/")

    g.bind("foaf", FOAF3)

    assert list(g.namespaces()) == [
        ("foaf", foaf2_uri),
        ("foaf1", foaf1_uri),
        ("foaf2", foaf3_uri),
    ]

    foaf4_uri = URIRef("http://xmlns.com/foaf/4.0/")
    FOAF4 = Namespace("http://xmlns.com/foaf/4.0/")

    g.bind("foaf", FOAF4)

    assert list(g.namespaces()) == [
        ("foaf", foaf2_uri),
        ("foaf1", foaf1_uri),
        ("foaf2", foaf3_uri),
        ("foaf3", foaf4_uri),
    ]
```

Phew! Sorry for taking so long to build a solid mental model.

Reply:
@gjhiggins  should these tests be added here? If so I will make a PR to add them, and see if they pass and if not discuss it with you.

Reply:
I guess so, ideally they should go into a file in the `test/test_namespace` folder but that folder doesn't yet exist in this branch.

I've (hopefully, helpfully) decanted the trimmed-down test code into a putative `test_namespace_rebinding.py` in [this gist](https://gist.github.com/gjhiggins/b4b59bffc18aa080ce466b01eb0e5d70).

Without the patch (posted above) the `test_binding_replace` (marked xfail) will xfail but will xpass if the patch is applied.

The other xfail pertains to the auto-creation of a namespace from a URIRef predicate.

Reply:
I made a PR with @gjhiggins changes here https://github.com/RDFLib/rdflib/pull/1800.

This is basically what you put in the comment here @gjhiggins - doctests should be fixed ideally  but I think our primary means of testing should always be unit tests, and doctests should be seen as mainly a mechanism to ensure that code in docs is correct.

I think it makes sense to merge https://github.com/RDFLib/rdflib/pull/1800 into this branch as is, as it fixes a bug, though I think the same issue occurs in other stores, so they should also be tested and fixed ideally. I will make another branch to do that.

Ideally this issue should be addressed in another branch, as we are putting a lot of eggs in one basket and basically playing on hard mode, but I guess the best we can do now is just push through and try and get things in from here.